### ISSUE TITLE
[no-unsafe-assignment] [no-unsafe-call] [no-unsafe-return] Improve user experience when the “real” issue is caught by TypeScript

### ISSUE THREAD

**Issue by lydell** [typescript-eslint/typescript-eslint#2665]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unsafe-assignment": "error",
    "@typescript-eslint/no-unsafe-call": "error",
    "@typescript-eslint/no-unsafe-return": "error",
  }
}
```

```TS
function run(tools: Array<string>): number {
  const tool = findMatchingTool(tools, "1.0")
  return getScore(tool);
}
```

```json
{
  "compilerOptions": {
    "allowSyntheticDefaultImports": true,
    "forceConsistentCasingInFileNames": true,
    "importsNotUsedAsValues": "error",
    "isolatedModules": true,
    "module": "CommonJS",
    "moduleResolution": "node",
    "noFallthroughCasesInSwitch": true,
    "noImplicitReturns": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "skipLibCheck": true,
    "strict": true,
    "target": "ES2019",
    "lib": [
      "ES2019",
    ],
    "outDir": "build"
  }
}
```

**Expected Result**

`const tool =` line: I only want TypeScript to say that `findMatchingTool` isn’t defined.

`return` line: I only want TypeScript to say that `getScore` isn’t defined.

**Actual Result**

ESLint:

```
/Users/lydell/test-project/tools.ts
  2:9   error  Unsafe assignment of an any value    @typescript-eslint/no-unsafe-assignment
  2:16  error  Unsafe call of an any typed value    @typescript-eslint/no-unsafe-call
  3:3   error  Unsafe return of an any typed value  @typescript-eslint/no-unsafe-return
  3:10  error  Unsafe call of an any typed value    @typescript-eslint/no-unsafe-call
```

TypeScript:

```
tools.ts:2:16 - error TS2304: Cannot find name 'findMatchingTool'.

2   const tool = findMatchingTool(tools, "1.0");
                 ~~~~~~~~~~~~~~~~

tools.ts:3:10 - error TS2304: Cannot find name 'getScore'.

3   return getScore(tool);
           ~~~~~~~~
```

The TypeScript errors are enough! Once they are fixed the `no-unsafe-*` errors are likely fixed as well!

**Additional Info**

Although I showed CLI output above for brevity, my issue is about the experience in VSCode (or any editor with ESLint support really).

I often write code that calls nonexistent functions. Once I’ve sketched out the code a bit I go back and try to actually implement those functions. I do this by jumping to the next error – sometimes I end up fixing something else, like a typo, and eventually I come across one of those functions that don’t exist yet. At first I might not remember that the function does not exist, so I read the error messages to find out what’s wrong. That can be a bit noisy. The _real_ error is that the function does not exist, but it gets lost in a sea of “unsafe any” errors.

When I see those “unsafe any” errors I think like “Duh, of course they’d be `any`, the function does not even exist”.

I think it would be awesome if the `no-unsafe-*` rules could avoid causing errors if there’s another TypeScript error causing them in the first place!

My goal is to improve the editing experience by reducing unnecessary errors/warnings that clutter the editor. If you have a better idea on how to do that I’d be all ears!

Possibly related: https://github.com/typescript-eslint/typescript-eslint/issues/2164

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.4.0` |
| `@typescript-eslint/parser`        | `4.4.0` |
| `TypeScript`                       | `4.0.3` |
| `ESLint`                           | `7.11.0` |
| `node`                             | `14.10.1` |

**Comment by bradzacher** [comment]
This is not really feasible to solve.

TypeScript doesn't tell us that this thing resolves to nothing - instead it just tells us its type is `any`.

The `any`s are special "error" `any`s, but they just tell us there was an error resolving the type - not what the error was. We can't just ignore these either, because often these error `any`s indicate a configuration mismatch (I.e. Our ts instance wasn't able to find a type but your IDE/tsc instance was).

In order to figure out if an `any` is actually a "not found" error, we'd have to ask typescript to compute the diagnostics.

Computing the diagnostics can take an average of 50ms per file. That might not sound like much, but that adds up to 1 second per 20 files! Across an entire project this kills lint times (I learned this the hard way with the `no-unused-vars-experimental` rule).

**Comment by lydell** [comment]
Sad to hear!

TypeScript + ESLint in VSCode is such a super nice way to write code. It’s a shame that they don’t feel more integrated, especially in the editor (on the command line one can just run `tsc && eslint .`).

On the same topic: https://github.com/microsoft/vscode-eslint/issues/1087.

Is there any heuristic we could use to catch maybe 80% of the cases? For example, for `no-unsafe-call` – if the called thing is a variable name and that name isn’t in scope then we could avoid creating an error?

**Comment by bradzacher** [comment]
Yeah unfortunately there's a small issue with how IDEs are designed now a days.

Everything is an extension, and everything is separate.

TS support is provided using (an inbuilt) extension. ESLint support is provided using (a 3rd party) extension.
These two extensions run separately and cannot share anything.

In an ideal world we would be able to talk to the TS language server that's providing types to the IDE, and build on top of it to save time and effort.
In that world we would be able to short-circuit the lint run entirely when TS reports a parsing error.
In that world we would be able to consume diagnostics for free, as they've ready been computed for the IDE.

But alas... That's not how it works (for various reasons like security).

There is actually some work that someone else has done that might be of interest to you. It integrates eslint into typescript as a language server plugin.
https://github.com/Quramy/typescript-eslint-language-service

So why not promote this as the defacto way to use our tooling? Mainly because the ESLint extension is more prevelant and requires less config to get working.

---

> Is there any heuristic we could use to catch maybe 80% of the cases? For example, for no-unsafe-call – if the called thing is a variable name and that name isn’t in scope then we could avoid creating an error?

The problem is that we don't know if you were referring to a global thing that doesn't exist because you've misconfigured something.

There's probably something we can do, but my intuition makes me thing that it's likely that most (if not all) cases of false-positives have a similar misconfiguration possibility.

With everything to do with linting, I always favor false-positives to false-negatives. False positives are noise you can ignore (annoying people is bad sure, but better than...). False negatives unknowingly introduce bugs and erode the user's trust in the ecosystem.

---

It's currently possible for us to detect error `any`s using private properties on the type, but that's all we can tell - that it's an error `any`.

There's probably more TS could report here without requiring us to reach for diagnostics, but I just haven't invested too much time into thinking about it. It'd need some thinking to provide a decent proposal to the TS team, and possibly more time to implement into TS itself.

Time is something I don't have enough of unfortunately 😅

**Comment by lydell** [comment]
Thanks for the detailed reply!

I think the “correct” solution to this problem is more `strict` options in TypeScript itself. It’s a bit odd IMO that you need a separate tool to find lurking `any` when there’s `noImplicityAny`. The rules that overlap with the compiler should be in the compiler. Then the separation between tools isn’t so bad anymore.

I’m currently diving into the TypeScript issue tracker to see if something like this has been discussed before.

**Comment by lydell** [comment]
I think we’re doomed:

- https://github.com/microsoft/TypeScript/pull/24423
- https://github.com/microsoft/TypeScript/issues/24737
- https://github.com/microsoft/TypeScript/issues/22716
- https://github.com/microsoft/TypeScript/issues/16296

Feel free to close the issue if it helps maintenance on this repo, or keep it open if you feel like it’s a legit thing to improve although unlikely to be fixed.

**Comment by ark120202** [comment]
Because of https://gist.github.com/RyanCavanaugh/f80f9ddc50d45c4d76e7c4101efada28 it's is actually quite likely to be added. There is https://github.com/microsoft/TypeScript/pull/40178, though for now it covers only property access.

**Comment by bradzacher** [comment]
As there's nothing we can strictly action in this project right now - I'm going to close this.
